File: packages/cactus-sync-client/lib/src/abstract/CactusModel.ts
import { QueryResult } from '@apollo/client'
import {
  ApolloQueryResult,
  FetchResult,
  OperationVariables,
} from '@apollo/client/core'
import { Version } from 'dexie'
import {
  DocumentNode,
  ExecutionResult,
  GraphQLFieldMap,
  GraphQLObjectType,
} from 'graphql'
import {
  AllGqlOperationsType,
  DefaultGqlOperations,
  DefaultGqlOperationType,
  getDefaultGqlOperations,
  SubscribeGqlOperationType,
} from '../graphql/DefaultGqlOperations'
import {
  ApolloRunner,
  ApolloRunnerExecute,
  ApolloSubscription,
} from './ApolloRunner'
import { Maybe } from './BasicTypes'
import { CactusSync } from './CactusSync'
import {
  notifyStateModelListeners,
  validateStateModelResult,
} from './StateModel'

interface CactusModelInitI {
  graphqlModelType: Maybe<GraphQLObjectType>
  defaultModelFragment?: Maybe<DocumentNode>
  // TODO: upgrade for versions like hooks?
  upgrade?(upgrade: Version['upgrade']): void
}

interface CactusModelDbInitI {
  db: CactusSync
  dbVersion: number
}

export interface CactusModelI extends CactusModelInitI, CactusModelDbInitI {}

export type CactusModelBuilder<
  TModel,
  TCreateInput = OperationVariables,
  TCreateResult = FetchResult<TModel>,
  TUpdateInput = OperationVariables,
  TUpdateResult = FetchResult<TModel>,
  TDeleteInput = OperationVariables,
  TDeleteResult = FetchResult<TModel>,
  TGetInput = OperationVariables,
  TGetResult = ApolloQueryResult<TModel>,
  TFindInput = OperationVariables,
  TFindResult = ApolloQueryResult<TModel>
> = (
  arg: CactusModelDbInitI
) => CactusModel<
  TModel,
  TCreateInput,
  TCreateResult,
  TUpdateInput,
  TUpdateResult,
  TDeleteInput,
  TDeleteResult,
  TGetInput,
  TGetResult,
  TFindInput,
  TFindResult
>

type OperationFunctionGql = {
  stringGql?: Maybe<string>
  fragmentGql?: Maybe<DocumentNode>
}

type OperationInput<TInput> = {
  input: TInput
  gql?: Maybe<OperationFunctionGql>
}

export type OperationFunction<TInput, TResult> = (
  input: TInput,
  gql?: Maybe<OperationFunctionGql>,
  notifyListeners?: Maybe<boolean>
) => Promise<ExecutionResult<TResult>>

export type QueryOperationFunction<
  TData = any,
  TFilter = OperationVariables,
  TResult = QueryResult<TData, TFilter>
> = (
  arg?: Maybe<TFilter>,
  gql?: Maybe<OperationFunctionGql>
) => Promise<ExecutionResult<TResult>>

export class CactusModel<
  TModel = any,
  TCreateInput = OperationVariables,
  TCreateResult = FetchResult<TModel>,
  TUpdateInput = OperationVariables,
  TUpdateResult = FetchResult<TModel>,
  TDeleteInput = OperationVariables,
  TDeleteResult = FetchResult<TModel>,
  TGetInput = OperationVariables,
  TGetResult = ApolloQueryResult<TModel>,
  TFindInput = OperationVariables,
  TFindResult = ApolloQueryResult<TModel>
> {
  modelName: string
  protected _defaultGqlOperations: DefaultGqlOperations
  protected _modelFields: (keyof TModel)[]
  defaultModelFragment?: Maybe<DocumentNode>
  db: CactusSync
  protected _graphqlModelType: GraphQLObjectType
  /**
   * We will remove any relationships by default for safety
   * User anyway in anytime may call it with custom gql
   * @param fields
   * @returns
   */
  protected _getModelFieldNames(fields: GraphQLFieldMap<any, any>) {
    return Object.values(fields)
      .filter(
        (el) =>
          !el.description?.includes('manyToOne') &&
          !el.description?.includes('oneToMany') &&
          !el.description?.includes('oneToOne')
      )
      .map((el) => el.name) as (keyof TModel)[]
  }
  constructor({
    graphqlModelType,
    db,
    upgrade,
    dbVersion,
    defaultModelFragment,
  }: CactusModelI) {
    if (graphqlModelType == null)
      throw Error(
        'graphqlModelType for CactusModel is not defined. Check type that you put in init functon'
      )
    this.defaultModelFragment = defaultModelFragment
    const fields = graphqlModelType.getFields()
    if (fields == null)
      throw Error(`no fields defined for ${graphqlModelType.name} model`)
    this._modelFields = this._getModelFieldNames(fields)
    this.db = db
    this._graphqlModelType = graphqlModelType
    this.modelName = graphqlModelType.name
    this._defaultGqlOperations = getDefaultGqlOperations({
      modelFields: this._modelFields,
      modelName: this.modelName,
    })
    if (upgrade) upgrade(db.version(dbVersion).upgrade)
  }

  static init<
    TModel,
    TCreateInput,
    TCreateResult,
    TUpdateInput,
    TUpdateResult,
    TDeleteInput,
    TDeleteResult,
    TGetInput,
    TGetResult,
    TFindInput,
    TFindResult
  >(
    arg: CactusModelInitI
  ): CactusModelBuilder<
    TModel,
    TCreateInput,
    TCreateResult,
    TUpdateInput,
    TUpdateResult,
    TDeleteInput,
    TDeleteResult,
    TGetInput,
    TGetResult,
    TFindInput,
    TFindResult
  > {
    return (dbInit: CactusModelDbInitI) =>
      new CactusModel({ ...arg, ...dbInit })
  }
  protected _graphqlRunner(): ApolloRunner<any> {
    const runner = this.db.graphqlRunner
    if (runner == null)
      throw Error(
        `Graphql runner is not defined! 
        Try to run "CactusSync.init()" in your app root
        before "CactusSync.createModel"`
      )
    return runner
  }
  protected async _execute<TVariables, TResult>(
    arg: ApolloRunnerExecute<TVariables>
  ) {
    return await this._graphqlRunner().execute<TModel, TVariables, TResult>(arg)
  }

  protected _resolveGql({
    fragmentGql,
    stringGql,
    operationType,
  }: {
    operationType: AllGqlOperationsType
    fragmentGql?: Maybe<DocumentNode>
    stringGql?: Maybe<string>
  }) {
    if (stringGql) return stringGql
    const fragment = fragmentGql ?? this.defaultModelFragment
    if (fragment) {
      return getDefaultGqlOperations({
        modelName: this.modelName,
        modelFragment: fragment,
      })[operationType]
    }
    return this._defaultGqlOperations[operationType]
  }
  protected async _executeMiddleware<TInput, TResult>(arg: {
    operationInput: OperationInput<TInput>
    operationType: DefaultGqlOperationType
    notifyListeners?: Maybe<boolean>
  }) {
    const { operationType, operationInput, notifyListeners } = arg
    const { input: variableValues, gql } = operationInput
    /**
     * If we receive fragmentGql, we concat it with default query
     * If we receive stringGql we replace default by stringGql
     * If class has default fragment it will be use it
     * And then it will be use default fields
     */
    const query = this._resolveGql({
      operationType,
      fragmentGql: gql?.fragmentGql,
      stringGql: gql?.stringGql,
    })
    const result = await this._execute<TInput, TResult>({
      variableValues,
      query,
      operationType,
    })

    /// STATE UDPATES

    if (!notifyListeners) return result
    const validateAndEmit = ({ remove }: { remove?: Maybe<boolean> }) => {
      const { isNotValid, data } = validateStateModelResult(result)
      console.log({ isNotValid, data })
      if (isNotValid || data == null) return
      for (const maybeModel of Object.values(data)) {
        notifyStateModelListeners({
          emitter: this.db.graphqlRunner.subscriptionsEmitter,
          remove: remove,
          item: maybeModel,
          notifyListeners,
          modelName: this.modelName,
        })
      }
    }
    switch (operationType) {
      case DefaultGqlOperationType.create:
      case DefaultGqlOperationType.update:
        validateAndEmit({})
        break
      case DefaultGqlOperationType.remove:
        validateAndEmit({ remove: true })
        break
    }
    return result
  }
  /**
   * By default this function will notify all states for this model
   * You can turn off it by turning notifyListeners = false
   *
   * @param input
   * @param gql
   * @param notifyListeners
   * @returns
   */
  add: OperationFunction<TCreateInput, TCreateResult> = async (
    input,
    gql,
    notifyListeners
  ) => {
    return await this._executeMiddleware({
      operationInput: {
        gql,
        input,
      },
      operationType: DefaultGqlOperationType.create,
      notifyListeners: notifyListeners == false ? false : true,
    })
  }
  /**
   * By default this function will notify all states for this model
   * You can turn off it by turning notifyListeners = false
   *
   * @param input
   * @param gql
   * @param notifyListeners
   * @returns
   */

  update: OperationFunction<TUpdateInput, TUpdateResult> = async (
    input,
    gql,
    notifyListeners
  ) => {
    return await this._executeMiddleware({
      operationInput: {
        gql,
        input,
      },
      operationType: DefaultGqlOperationType.update,
      notifyListeners: notifyListeners == false ? false : true,
    })
  }
  /**
   * By default this function will notify all states for this model
   * You can turn off it by turning notifyListeners = false
   *
   * @param input
   * @param gql
   * @param notifyListeners
   * @returns
   */

  remove: OperationFunction<TDeleteInput, TDeleteResult> = async (
    input,
    gql,
    notifyListeners
  ) => {
    return await this._executeMiddleware({
      operationInput: {
        gql,
        input,
      },
      operationType: DefaultGqlOperationType.remove,
      notifyListeners: notifyListeners == false ? false : true,
    })
  }
  get: OperationFunction<TGetInput, TGetResult> = async (input, gql) => {
    return await this._executeMiddleware({
      operationInput: {
        gql,
        input,
      },
      operationType: DefaultGqlOperationType.get,
    })
  }
  find: QueryOperationFunction<TModel, TFindInput, TFindResult> = async (
    arg,
    gql
  ) => {
    return await this._executeMiddleware({
      operationInput: {
        gql,
        input: arg ?? undefined,
      },
      operationType: DefaultGqlOperationType.find,
    })
  }

  /// ================= Replication section ========================
  protected get _allReplicationQueries() {
    const queries = Object.values(SubscribeGqlOperationType).map((type) =>
      this._resolveGql({
        operationType: type,
      })
    )
    return queries
  }
  async startReplication() {
    return await this.db.startModelReplication({
      modelName: this.modelName,
      queries: this._allReplicationQueries,
    })
  }
  async stopReplication() {
    return await this.db.startModelReplication({
      modelName: this.modelName,
      queries: this._allReplicationQueries,
    })
  }
  get isReplicating(): boolean {
    return this.db.isModelReplicating({ modelName: this.modelName })
  }
  get graphqlSubscriptions(): Maybe<
    ApolloSubscription<TCreateResult | TUpdateResult | TDeleteResult>
  >[] {
    const subscriptions =
      this.db.graphqlRunner?.getModelSubscriptions({
        modelName: this.modelName,
      }) ?? []
    console.log('Cactus model - graphqlSubscriptions', { subscriptions })

    return subscriptions
  }
}

File: dart_packages/cactus_sync_client/lib/src/abstract/cactus_model.dart
import 'package:graphql/client.dart';

import '../graphql/gql_builder.dart';
import '../graphql/graphql_result.dart';
import '../utils/utils.dart';
import 'cactus_sync.dart';
import 'graphql_runner.dart';
import 'recorded_model.dart';

/// [stringQueryGql] is a gql which replaces the whole gql
// TODO(arenukvern): add an example
/// [modelFragmentGql] is used to fill requested fields only in gql
// TODO(arenukvern): add an example
class QueryGql {
  const QueryGql({
    final this.stringQueryGql,
    final this.modelFragmentGql,
  });
  final String? stringQueryGql;
  final String? modelFragmentGql;
}

typedef CactusModelBuilder<
        TModel,
        TCreateInput extends JsonSerializable,
        TCreateResult,
        TUpdateInput extends JsonSerializable,
        TUpdateResult,
        TDeleteInput extends JsonSerializable,
        TDeleteResult,
        TGetResult,
        TFindInput extends JsonSerializable,
        TFindResult>
    = CactusModel<
            TModel,
            TCreateInput,
            TCreateResult,
            TUpdateInput,
            TUpdateResult,
            TDeleteInput,
            TDeleteResult,
            TGetResult,
            TFindInput,
            TFindResult>
        Function({required CactusSync db});

/// Abstract Model class to insure consistency in CUDGF
abstract class AbstractCactusModel<
    TCreateInput,
    TCreateResult,
    TUpdateInput,
    TUpdateResult,
    TRemoveInput,
    TRemoveResult,
    TGetResult,
    TFindInput,
    TFindResult> {
  Future<GraphqlResult<TCreateResult>> create({
    required final TCreateInput variableValues,
    final QueryGql? queryGql,
    final bool notifyListeners = true,
  });
  Future<GraphqlResult<TUpdateResult>> update({
    required final TUpdateInput variableValues,
    final QueryGql? queryGql,
    final bool notifyListeners = true,
  });
  Future<GraphqlResult<TRemoveResult>> remove({
    required final TRemoveInput variableValues,
    final QueryGql? queryGql,
    final bool notifyListeners = true,
  });
  Future<GraphqlResult<TGetResult>> get({
    required final RecordedModel variableValues,
    final QueryGql? queryGql,
    final bool notifyListeners = true,
  });
  Future<GraphqlResult<TFindResult>> find({
    required final TFindInput variableValues,
    final QueryGql? queryGql,
    final bool notifyListeners = true,
  });
}

class CactusModel<
        TType,
        TCreateInput extends JsonSerializable,
        TCreateResult,
        TUpdateInput extends JsonSerializable,
        TUpdateResult,
        TRemoveInput extends JsonSerializable,
        TRemoveResult,
        TGetResult,
        TFindInput extends JsonSerializable,
        TFindResult>
    implements
        AbstractCactusModel<
            TCreateInput,
            TCreateResult,
            TUpdateInput,
            TUpdateResult,
            TRemoveInput,
            TRemoveResult,
            TGetResult,
            TFindInput,
            TFindResult> {
  CactusModel({
    required final this.createFromJsonCallback,
    required final this.updateFromJsonCallback,
    required final this.removeFromJsonCallback,
    required final this.getFromJsonCallback,
    required final this.findFromJsonCallback,
    required final this.defaultModelFragment,
    required final this.graphqlModelFieldNames,
    required final this.graphqlModelName,
    required final this.db,
  }) {
    gqlBuilder = GqlBuilder(
      modelName: graphqlModelName,
      modelFields: graphqlModelFieldNames,
      modelFragment: defaultModelFragment,
    );
  }
  FromJsonCallback<TCreateResult> createFromJsonCallback;
  FromJsonCallback<TFindResult> findFromJsonCallback;
  FromJsonCallback<TGetResult> getFromJsonCallback;
  FromJsonCallback<TRemoveResult> removeFromJsonCallback;
  FromJsonCallback<TUpdateResult> updateFromJsonCallback;

  final CactusSync db;
  final String graphqlModelName;
  late final GqlBuilder gqlBuilder;
  final List<String?> graphqlModelFieldNames;
  final String defaultModelFragment;

  static CactusModelBuilder<
      TModel,
      TCreateInput,
      TCreateResult,
      TUpdateInput,
      TUpdateResult,
      TDeleteInput,
      TDeleteResult,
      TGetResult,
      TFindInput,
      TFindResult> init<
          TModel,
          TCreateInput extends JsonSerializable,
          TCreateResult,
          TUpdateInput extends JsonSerializable,
          TUpdateResult,
          TDeleteInput extends JsonSerializable,
          TDeleteResult,
          TGetResult,
          TFindInput extends JsonSerializable,
          TFindResult>({
    required final List<String?> graphqlModelFieldNames,
    required final String graphqlModelName,
    required final String defaultModelFragment,
    required final FromJsonCallback<TCreateResult> createFromJsonCallback,
    required final FromJsonCallback<TFindResult> findFromJsonCallback,
    required final FromJsonCallback<TGetResult> getFromJsonCallback,
    required final FromJsonCallback<TDeleteResult> removeFromJsonCallback,
    required final FromJsonCallback<TUpdateResult> updateFromJsonCallback,
  }) =>
      ({required final db}) => CactusModel(
            createFromJsonCallback: createFromJsonCallback,
            db: db,
            defaultModelFragment: defaultModelFragment,
            findFromJsonCallback: findFromJsonCallback,
            getFromJsonCallback: getFromJsonCallback,
            graphqlModelFieldNames: graphqlModelFieldNames,
            graphqlModelName: graphqlModelName,
            removeFromJsonCallback: removeFromJsonCallback,
            updateFromJsonCallback: updateFromJsonCallback,
          );

  FromJsonCallback<TQueryResult>
      _getFromJsonCallbackByOperationType<TQueryResult>({
    required final DefaultGqlOperationType operationType,
  }) {
    final callback = (() {
      switch (operationType) {
        case DefaultGqlOperationType.create:
          return createFromJsonCallback;
        case DefaultGqlOperationType.update:
          return updateFromJsonCallback;
        case DefaultGqlOperationType.remove:
          return removeFromJsonCallback;
        case DefaultGqlOperationType.get:
          return getFromJsonCallback;
        case DefaultGqlOperationType.find:
          return findFromJsonCallback;
        case DefaultGqlOperationType.fromString:
          throw Exception(
            'DefaultGqlOperationType is fromString but '
            'has to be different',
          );
      }
    })();
    return callback as FromJsonCallback<TQueryResult>;
  }

  GraphqlRunner? get _graphqlRunner => db.graphqlRunner;
  Future<GraphqlResult<TQueryResult>>
      _execute<TVariables extends JsonSerializable, TQueryResult>({
    required final String query,
    required final TVariables variableValues,
    required final DefaultGqlOperationType operationType,
    required final FromJsonCallback<TQueryResult> fromJsonCallback,
  }) async {
    final resolvedGraphqlRunner = _graphqlRunner;
    if (resolvedGraphqlRunner == null) {
      throw ArgumentError.notNull('_graphqlRunner');
    }
    CactusSync.l.info({'is graphqlRunner initialized'});
    return resolvedGraphqlRunner.execute<TVariables, TQueryResult>(
      fromJsonCallback: fromJsonCallback,
      operationType: operationType,
      query: query,
      variableValues: variableValues,
    );
  }

  String _resolveOperationGql({
    required final DefaultGqlOperationType operationType,
    final QueryGql? queryGql,
  }) {
    final stringQueryGql = queryGql?.stringQueryGql;
    if (stringQueryGql != null && stringQueryGql.isNotEmpty) {
      return stringQueryGql;
    }
    final modelFragmentGql = queryGql?.modelFragmentGql;
    final builder = modelFragmentGql != null && modelFragmentGql.isNotEmpty
        ? GqlBuilder(
            modelName: graphqlModelName,
            modelFragment: modelFragmentGql,
          )
        : gqlBuilder;
    return builder.getByOperationType(operationType: operationType);
  }

  Future<GraphqlResult<TResult>>
      _executeMiddleware<TVariables extends JsonSerializable, TResult>({
    required final DefaultGqlOperationType operationType,
    required final TVariables variableValues,
    final QueryGql? queryGql,
    final bool notifyListeners = true,
  }) async {
    /**
     * If we receive modelFragmentGql, we concat it with default query
     * If we receive stringQueryGql we replace default by stringQueryGql
     * If class has default fragment it will be use it
     * And then it will be use default fields
     */
    final query = _resolveOperationGql(
      operationType: operationType,
      queryGql: queryGql,
    );
    final fromJsonCallback = _getFromJsonCallbackByOperationType<TResult>(
      operationType: operationType,
    );
    CactusSync.l.info({
      'query': query,
      'jsonCallback': fromJsonCallback,
    });

    final result = await _execute<TVariables, TResult>(
      variableValues: variableValues,
      query: query,
      operationType: operationType,
      fromJsonCallback: fromJsonCallback,
    );

    CactusSync.l.info(result);

    /// STATE UPDATES
    return result;
    // TODO: enable sync for different states
    // if (!notifyListeners) return result;
    // var validateAndEmit = ({bool? remove}){
    //   var { isNotValid, data } = validateStateModelResult(result);
    //   if (isNotValid || data == null) return;
    //   for (var maybeModel of Object.values(data)) {
    //     notifyStateModelListeners({
    //       emitter: this.db.graphqlRunner.subscriptionsEmitter,
    //       remove: remove,
    //       item: maybeModel,
    //       notifyListeners,
    //       modelName: this.modelName,
    //     });
    //   }
    // }
    // switch (operationType) {
    //   case DefaultGqlOperationType.create:
    //   case DefaultGqlOperationType.update:
    //     validateAndEmit({});
    //     break
    //   case DefaultGqlOperationType.remove:
    //     validateAndEmit({ remove: true });
    //     break
    // }
    // return result;
  }

  @override
  Future<GraphqlResult<TCreateResult>> create({
    required final TCreateInput variableValues,
    final QueryGql? queryGql,
    final bool notifyListeners = true,
  }) async {
    CactusSync.l.info('create');
    return _executeMiddleware(
      operationType: DefaultGqlOperationType.create,
      queryGql: queryGql,
      notifyListeners: notifyListeners,
      variableValues: variableValues,
    );
  }

  @override
  Future<GraphqlResult<TUpdateResult>> update({
    required final TUpdateInput variableValues,
    final QueryGql? queryGql,
    final bool notifyListeners = true,
  }) async {
    CactusSync.l.info('update');
    return _executeMiddleware(
      operationType: DefaultGqlOperationType.update,
      queryGql: queryGql,
      notifyListeners: notifyListeners,
      variableValues: variableValues,
    );
  }

  @override
  Future<GraphqlResult<TRemoveResult>> remove({
    required final TRemoveInput variableValues,
    final QueryGql? queryGql,
    final bool notifyListeners = true,
  }) async {
    CactusSync.l.info('remove');
    return _executeMiddleware(
      operationType: DefaultGqlOperationType.remove,
      queryGql: queryGql,
      notifyListeners: notifyListeners,
      variableValues: variableValues,
    );
  }

  @override
  Future<GraphqlResult<TFindResult>> find({
    required final TFindInput variableValues,
    final QueryGql? queryGql,
    final bool notifyListeners = true,
  }) async {
    CactusSync.l.info('find');
    return _executeMiddleware(
      operationType: DefaultGqlOperationType.find,
      queryGql: queryGql,
      notifyListeners: notifyListeners,
      variableValues: variableValues,
    );
  }

  @override
  Future<GraphqlResult<TGetResult>> get({
    required final RecordedModel variableValues,
    final QueryGql? queryGql,
    final bool notifyListeners = true,
  }) async {
    CactusSync.l.info('get');
    return _executeMiddleware(
      operationType: DefaultGqlOperationType.get,
      queryGql: queryGql,
      notifyListeners: notifyListeners,
      variableValues: variableValues,
    );
  }
}
